text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+to create their own extension: `MyServiceCustomizer` which modifies the underlying L2
+construct properties directly.
+
+Another drawback of this approach is that it creates a dependence on the L3 construct.
+This `Service` class implements attractive new features that can not be easily implemented
+at L2 without deep knowledge of the feature itself.
+via these L3 construct extensions it will be a massive effort to reimplement the same
+results directly at the L2 level. Basically once you start implementing your service
+using this L3 construct it would be almost impossible to refactor your CDK app back into
+lower level L2 constructs because the surface area and logic that the L3 construct
+implements is very broad.
+
+1. __Do we anticipate certain extensions conflicting with each other?__
+
+   Yes it is possible that extensions may conflict with each other. It is the responsibility
+   of each extension to check for other extensions at the time of being added to the service
+   and throw an exception if there is an expected conflict.
+
+2. __Given that ECS patterns are meant to push best practices in application
+   infrastructure/architecture, how will we provide that guidance with extensions?__
+
+   The goal of these extensions is to do the right thing by default, similar to the
+   existing ECS patterns, but the user just has more choices about what combinations
+   of feature integrations they want to enable. When using the built-in extensions there
+   should be no "bad" choices. If a user wants to leave out an extension it is safe to
+   do so, and if they chose to enable it there should be no gotchas from adding another
+   extension. The goal is to make this hard to misuse, but still powerful and extendable.
+
+# Rationale and Alternatives
+
+One alternative approach to this is to try to enhance the lower level L2 methods with
+an extension type system. The existing
+[`TaskDefinition.addExtension()`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.TaskDefinition.html#add-wbr-extensionextension)
+method is the beginning of an attempt at this.
+
+However attempting to build out extensions on the L2 level will fail to fully meet customer
+expectations because the features they want to enable bridge multiple different
+types of L2 construct. For example in order for an extension like App Mesh to work
+properly it must change the service to have a cloud map configuration, change the
+task definition IAM role to have the right level of access, change the task definition
+to add proxy settings, attach a container to the task definition, and create and link
+App Mesh virtual service and virtual node to the service. Because the full feature
+requires touching all of these different constructs and resulting resources it is not
+possible to fully implement an extension system at the L2 level. For this reason the
+existing `TaskDefinition.addExtension()` is not a viable alternative.
+
```

Comment:
From a design perspective, have you considered the following alternative:

1. Define extensions as simple constructs instead of introducing an additional composition model. Constructs should be sufficient.
2. Expose read/write APIs in `Service` so that extensions can interact with it.

The result would look something like this, which includes less moving parts and is more idiomatic to CDK:

```ts
const svc = new Service(this, 'MyService', { environment });

new CloudWatchAgentExtension(this, 'CloudWatchAgent', { service: svc, ... });
new AppMeshExtension(this, 'AppMesh', { service: svc, ... });
new FireLensExtension(...);
```

Another alternative that you could consider is a subclass-based API. In this type of API, the extensions will "discover" the service by traversing up the construct tree (similar to how the `Stack` is discovered). I am not sure that's a superior model.

```ts
class MyAwesomeService extends Service {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new CloudWatchAgent(this, 'Foo', { ... });
}
```

Reply:
In this case the limitation is the underlying L2 constructs, and the approach to immutability.

So far from what I've seen in CDK the vast majority of the properties of a CDK construct are always readonly: they get set once at constructor time and are immutable from that point forward. This L3 API works within that system. The extensions in the `ServiceDescrpition` define the properties for the underlying L2 construct, and then when you create that L3 construct it actually creates the underlying L2 constructs, which create the underlying `Cfn` resources, baking everything into an immutable state from that point forward.

So the general idea is that this will stick with customer expectations that once they call `new Construct(stack, 'name', props)` the underlying construct is immutable from that point forward. Extensions can't mutate other constructs that are already created. (I don't think I've seen anything similar to that anywhere else in CDK, I'd be happy to learn from an existing example of this if there is one that I'm unaware of).

The idea of a subclass is interesting. However it is a lot more boilerplate to define a service, and its not as intuitive of an interface as just saying `Service.add(Feature)`